R/score_forecasts.R

```diff
@@ -153,105 +155,100 @@ score_forecasts <- function(
-          overprediction = rowSums(dplyr::select(., dplyr::starts_with("overprediction")), na.rm = FALSE) / (n_interval_scores - 0.5 * (exists_interval_score_0)),
-          underprediction = rowSums(dplyr::select(., dplyr::starts_with("underprediction")), na.rm = FALSE) / (n_interval_scores - 0.5 * (exists_interval_score_0))
-        ) %>%
-        dplyr::select(
-          -dplyr::starts_with("aem_"),
-          -dplyr::starts_with("ae_point_"),
-          -dplyr::starts_with("interval_score"),
-          -dplyr::starts_with("sharpness_"),
-          -dplyr::starts_with("underprediction_"),
-          -dplyr::starts_with("overprediction_")
-        ) %>%
-        dplyr::rename(dispersion = sharpness)%>%
-        dplyr::select(
-          1:8, dplyr::starts_with("coverage_"),
-          dplyr::starts_with("abs_error"),
-          "n_interval_scores", "exists_interval_score_0", "wis",
-          "dispersion", "overprediction", "underprediction"
-        )
-    }
-  )
+      purrr::set_names(~ sub(abs_var_rename, "abs_error", .x)) %>%
+      ## need to remove all columns ending with NA to not affect WIS calculations
+      dplyr::select(
+        -dplyr::ends_with("_NA")
+      ) %>%
+      ## before next lines: do we need to check to ensure interval_score columns exist?
+      ## the following lines ensure that we use denominator for the wis of
+      ## (# of interval_scores)-0.5
+      ## which is written in the paper and elsewhere as
+      ## (# of interval_scores at level >0 ) + 0.5 or (K + 1/2)
+      ## to make sure that the median only gets half the weight of the other
+      ## intervals, multiply its value by 0.5
+      dplyr::mutate(
+        n_interval_scores = rowSums(!is.na(dplyr::select(., dplyr::starts_with("interval_score")))),
+        exists_interval_score_0 = "interval_score_0" %in% names(.),
+        interval_score_0 = ifelse(exists_interval_score_0, 0.5 * interval_score_0, NA_real_),
+        dispersion_0 = ifelse(exists_interval_score_0, 0.5 * dispersion_0, NA_real_),
+        underprediction_0 = ifelse(exists_interval_score_0, 0.5 * underprediction_0, NA_real_),
+        overprediction_0 = ifelse(exists_interval_score_0, 0.5 * overprediction_0, NA_real_)
+      ) %>%
+      dplyr::mutate(
+        wis = rowSums(dplyr::select(., dplyr::starts_with("interval_score")), na.rm = FALSE) / (n_interval_scores - 0.5 * (exists_interval_score_0)),
+      ) %>%
+      dplyr::mutate(
+        dispersion = rowSums(dplyr::select(., dplyr::starts_with("dispersion")), na.rm = FALSE) / (n_interval_scores - 0.5 * (exists_interval_score_0)),
+        overprediction = rowSums(dplyr::select(., dplyr::starts_with("overprediction")), na.rm = FALSE) / (n_interval_scores - 0.5 * (exists_interval_score_0)),
+        underprediction = rowSums(dplyr::select(., dplyr::starts_with("underprediction")), na.rm = FALSE) / (n_interval_scores - 0.5 * (exists_interval_score_0))
+      ) %>%
```

Comment:
I think we've discussed this before, but -- does `scoringutils` not calculate `wis`, `dispersion`, `underprediction`, and `overprediction` in a convenient way? It seems surprising that we have to do so much of this calculation by hand.